pandas/tests/series/test_analytics.py

```diff
@@ -237,6 +238,60 @@ def test_npdiff(self):
         r = np.diff(s)
         assert_series_equal(Series([nan, 0, 0, 0, nan]), r)
 
+    def test_diff(self):
+        # Combined datetime diff, normal diff and boolean diff test
+        ts = tm.makeTimeSeries(name="ts")
+        ts.diff()
+
+        # int dtype
+        a = 10000000000000000
+        b = a + 1
+        s = Series([a, b])
+
+        result = s.diff()
+        assert result[1] == 1
+
+        # neg n
+        result = ts.diff(-1)
+        expected = ts - ts.shift(-1)
+        assert_series_equal(result, expected)
+
+        # 0
+        result = ts.diff(0)
+        expected = ts - ts
+        assert_series_equal(result, expected)
+
+        # datetime diff (GH3100)
+        s = Series(date_range("20130102", periods=5))
+        result = s - s.shift(1)
+        expected = s.diff()
+        assert_series_equal(result, expected)
+
+        # timedelta diff
+        result = result - result.shift(1)  # previous result
+        expected = expected.diff()  # previously expected
+        assert_series_equal(result, expected)
+
+        # with tz
+        s = Series(
+            date_range("2000-01-01 09:00:00", periods=5, tz="US/Eastern"), name="foo"
+        )
+        result = s.diff()
+        expected = Series(TimedeltaIndex(["NaT"] + ["1 days"] * 4), name="foo")
+        assert_series_equal(result, expected)
+
+        # boolean series (test for fixing #17294)
+        s = Series([False, True, True, False, False])
+        result = s.diff()
+        expected = Series([nan, True, False, True, False])
+        assert_series_equal(result, expected)
+
+        # boolean nan series
```

Comment:
I don't agree that this should differ from the previous case. IMO, both of these are boolean dtype. Pandas just doesn't have a nullable boolean dtype (yet) but we generally try to treat a mix of bool and NA the same as bool.

So I think that this should also be object dtype, but the values should be booleans instead of integers. Is that difficult to support? I suppose it'll require a scan of the values for the `object`-dtype case, to ensure that everything is boolean or NA?

Reply:
When I previously tested my code I did notice a difference, hence why I added the tests as such. If you are sure that pandas does not treat it any different and as such does not require testing this now and in the future, I will gladly remove the first test.

In regards to your second point, I could definitely change the code to create only a boolean series instead, but why? The same could be done with logic instead of diff. I think the point of diff is to see the difference between the current value and the previous one (e.g. True - False = 1 and False - True = -1). This is also in line with the subtraction done by diff on other non-boolean series and what happens if you subtract booleans in Python. I found this bug looking for ways to detect these changes. Or is that not what you meant?

Reply:
I'm having a hard time following, but IMO these two should have the same output

```python
In [3]: pd.Series([True, False, True]).diff()
Out[3]:
0     NaN
1    True
2    True
dtype: object

In [4]: pd.Series([True, False, True]).astype(object).diff()
Out[4]:
0    NaN
1     -1
2      1
dtype: object
```

I think both of those should be `Out[3]`, the same result as `np.diff` with the missing values handled appropriately.

Reply:
I seem to be missing something, when I run np.diff on the series it results in this:

```
In [2]: np.diff(pd.Series([False, True]))
Out[2]: array([ True])

In [3]: np.diff(pd.Series([np.nan, False, True]))
Out[3]: array([nan, 1], dtype=object)
```
Which is pretty close to the current pandas implementation. Do we want to remove the first item in the list that currently results in np.nan? That would mean changing the other code in the diff function as well. Ignoring that, the results are the same as my code. As the dtypes are the problem, we could either convert the series before or do the same numpy does.

Reply:
Just to make sure, I also tested np.diff on a numpy array with the same results:

```
In [4]: np.diff(np.array([True, False, np.nan]))
Out[4]: array([-1., nan])

In [5]: np.diff(np.array([True, False]))
Out[5]: array([ True])
```

Reply:
I think we want the same behavior as `np.diff` on a boolean-dtype ndarray.

Because of the missing values, we'll always return an object-dtype Series where the values are either NaN, True, or False.

cc @jbrockmendel if you have thoughts.

Reply:
It sounds like this would be a lot easier if we had a BoolNA in place.  Is that accurate?  Since bool-dtype is a thin wrapper over uint8, I wonder how hard it would be to adapt IntegerNA to this purpose (@WillAyd you had a PR about this a while back, right?)

Reply:
Long time ago. @jreback had something more modern in #25415

Reply:
Yes, this would be easier with BoolNA.

I think we want to achieve the BoolNA behavior, but with object dtype today.

Reply:
am -1 on doing anything with object dtype as it makes the code much too complex.